app/helpers/posts_helper.rb

```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+module PostsHelper
+  include ApplicationHelper
+
+  def self.clear_canonical_cache!(post)
+    key = canonical_redis_key(post)
+    Discourse.redis.del(key)
+  end
+
+  def self.canonical_redis_key(post)
+    "post_canonical_url_#{post.id}"
+  end
+
+  def cached_post_url(post, use_canonical:)
+    if use_canonical
+      # this is very expensive, we cache it for 12 hours to ensure N+1 is not
+      key = PostsHelper.canonical_redis_key(post)
+
+      url = Discourse.redis.get(key)
+
+      # break cache if either slug or topic_id changes
+      if url && !url.start_with?(post.topic.url)
```

Comment:
I think it'll be useful if we can test cache breakage too.

Reply:
Its an expensive test, but I guess we can fiddle with ids to prove it works.  on the upside though the code is covered now. 

proving no expensive N+1 is at play is going to be way more expensive of a test and that is kind of the critical thing here.

Reply:
Perhaps a test that fetches a post and then we update the post's topic_id? Then we can assert that the url is no longer pointing at the old topic. That should be a pretty cheap test.